expan/core/binning.py

```diff
@@ -8,725 +6,248 @@
 
 from expan.core.util import is_number_and_nan
 
-symbol_universal_set = 'ξ'  # https://en.wikipedia.org/wiki/Xi_(letter)
 logger = logging.getLogger(__name__)
 
-class Binning(object):
-    def __init__(self):
-        """
-        The Binning class has two subclasses: CategoricalBinning and NumericalBinning.
-        """
-        pass
-
-    def __len__(self):
-        raise NotImplementedError('must implement this in subclass')
-
-    def label(self, data, format_str=None):
-        """
-        This returns the bin labels associated with each data point in the series,
-        essentially 'applying' the binning to data.
-
-        Args:
-            data (array-like): array of datapoints to be binned
-            format_str (str): string defining the format of the label to apply
-
-        Returns:
-            array-like: array of the bin label corresponding to each data point.
-
-        Note:
-            Implemented in subclass.
-        """
-        raise NotImplementedError('must implement this in subclass')
-
-    def __str__(self, format_str='{standard}'):
-        res = "{} with {:d} bins:".format(self.__class__.__name__, len(self))
-        for ii, lbl in enumerate(self.labels(format_str)):
-            res += '\n {:d}: {}'.format(ii, lbl)
-        return res
-
-
-class CategoricalBinning(Binning):
-    """
-    A CategoricalBinning is essentially a list of lists of categories. 
-    Each bin within a Binning is an ordered list of categories.
-    """
-
-    def __init__(self, data=None, nbins=None):
-        """CategoricalBinning constructor
 
-        Args:
-            data (array-like): array of datapoints to be binned
-            nbins (int): number of bins
-        """
-        super(CategoricalBinning, self).__init__()
-        if data is None:
-            # NB: this will go through the property setters, so they will not be simply lists.
-            self.categories = []
+#------- bin representations -------#
+class Bin(object):
+    # this is a necessary hack for the buggy implementation of assertItemsEqual in python2
+    # see https://stackoverflow.com/a/29690198
+    # note that if we only use python3, assertCountEqual in python3 solves this problem
+    __hash__ = None
+
+    def __init__(self, bin_type, *repr_args):
+        '''
+        Constructor for a bin object.
+        :param id: identifier (e.g. bin number) of the bin
+        :param bin_type: "numerical" or "categorical"
+        :param repr_args: arguments to represent this bin. 
+                          args for numerical bin includes lower, upper, lower_closed, upper_closed
+                          args for categorical bin includes a list of values
+        '''
+        if bin_type == "numerical" and len(repr_args) != 4:
```

Comment:
These conditions feels like adding two derived classes might be a good idea to consider.
For instance:
`class NumericalBin(Bin):
def __init__(self, lower, upper, lower_closed, upper_closed)
class CategoricalBin(Bin):
def __init__(self, categories)`

Reply:
do you mean use ```class NumericalBin(Bin)``` and get rid of ```NumericalRepresentation```?

Reply:
Well, yes, maybe in this way, like you say.
The point is, as far as I see the only responsibility of Bin class is to receive some parameters, and decide which binning it is.
But one can solve it on the level of class design.

Reply:
Feel free to postpone it, it is a stylistic comment, we can refactor it later.
[PS. However, "there is no time for refactoring"]